File: JAVA programs/Merge_Sort.java
import java.util.Scanner;

public class Merge_Sort {
    public static void main(String[]args)
    {
        Scanner S=new Scanner(System.in);
        int n;
        System.out.println("Enter the Length of the Array : ");
        n=S.nextInt();
        int[]arr=new int[n];
        System.out.println("Enter the  Array Elements : ");
        for(int i=0;i<n;i++)
        {
            System.out.print("Element "+(i+1)+" : ");
            arr[i]=S.nextInt();
        }
        display(arr);
        System.out.println("\nSorted Array Using Merge Sort ");
        Mergesort(arr,0,arr.length-1);
        display(arr);
    }
    static void display(int[]arr)
    {
        for(int i=0;i<arr.length;i++)
        {
            System.out.print("\t"+arr[i]);
        }
    }
    static void Mergesort(int[]arr,int l,int u)
    {
        if(l<u)
        {
            int mid=(l+u)/2;
            Mergesort(arr,l,mid);
            Mergesort(arr,mid+1,u);

            merge(arr,l,u,mid);
        }
    }
    static void merge(int[]arr,int l,int u,int mid)
    {
        int l1=mid-l+1;
        int l2=u-mid;
        int[]a1=new int[l1];
        int[]a2=new int[l2];
        for(int p=0;p<l1;p++)
        {
            a1[p]=arr[p+l];
        }
        for(int p=0;p<l2;p++)
        {
            a2[p]=arr[p+mid+1];
        }
        int i=0,j=0,k=l;
        while(i<l1 && j<l2)
        {
          if(a1[i]>a2[j])
          {
              arr[l]=a2[j];
              l++;j++;
          }
          else if (a1[i]<a2[j])
          {
              arr[l]=a1[i];
              l++;i++;
          }
          else {
              arr[l] = a1[i];
              l++;i++;
          }
        }
        while(i<l1)
        {
            arr[l]=a1[i];
            l++;i++;
        }
        while(j<l2)
        {
            arr[l]=a2[j];
            l++;
            j++;
        }
    }
}

File: C++/Merge_Sort.cpp
#include<iostream>
using namespace std;
void display(int arr[],int n)
{
   for(int i=0;i<n;i++)
   {
      cout<<" "<<arr[i]<<" ";
   }
}
void merge(int arr[],int s,int mid,int n)
{
   int l1=mid-s+1;
   int l2=n-mid;
   int a1[l1],a2[l2];

   for(int i=0;i<l1;i++)
   {
      a1[i]=arr[i+s];
   }
   for(int i=0;i<l2;i++)
   {
      a2[i]=arr[i+mid+1];
   }
   int p=0,q=0,r=s;
   while(p<l1 && q<l2)
   {
      if(a1[p]<a2[q])
      {
         arr[r]=a1[p];
         p++;
      }
      else{
         arr[r]=a2[q];
         q++;
      }
      r++;
   }
   while(p<l1)
   {
      arr[r]=a1[p];
      p++;
      r++;
   }
   while(q<l2)
   {
      arr[r]=a2[q];
      q++;
      r++;
   }
}
void merge_sort(int arr[],int s,int n)
{
  if(s<n)
   {
      int mid=(s+n)/2;

      merge_sort(arr,s,mid);
      merge_sort(arr,mid+1,n);

      merge(arr,s,mid,n);
   }
}
int main()
{
   int n;
   cout<<"Enter the Number : \n";
   cin>>n;
   int arr[n];
   cout<<"Enter the Array Elements : \n";
   for(int i=0;i<n;i++)
   {
      cout<<"Element "<<i+1<<endl;
      cin>>arr[i];
   }
   cout<<"The Array is "<<endl;
   display(arr,n);
   cout<<"The Sorted Array is : \n";
   merge_sort(arr,0,n);
   display(arr,n);
   return 0;
}